src/proxies/clientSDKCacheProxy.ts

```diff
@@ -0,0 +1,37 @@
+import { CogniteClient } from '@cognite/sdk';
+import { ClientSDKCache } from '../cache/ClientSDKCache';
+
+type CachedKeys = keyof ClientSDKCache;
+type SDKKeys = Exclude<keyof CogniteClient, 'project'>;
+type CachedTypes = ClientSDKCache[CachedKeys];
+type SDKTypes = CogniteClient[SDKKeys];
+
+export const wrapInCacheProxy = (
+  client: CogniteClient,
+  cache: ClientSDKCache
+): CogniteClient => {
+  const clientHandler: ProxyHandler<CogniteClient> = {
+    get(target, name: SDKKeys | CachedKeys) {
+      if (cache[name as keyof ClientSDKCache]) {
+        return new Proxy(
+          cache[name as CachedKeys],
+          apiHandler<CachedTypes, SDKTypes>(name as CachedKeys)
+        );
+      } else {
+        return target[name];
+      }
+    },
+  };
+
+  const apiHandler = <T extends CachedTypes, K extends SDKTypes>(
+    name: CachedKeys
+  ): ProxyHandler<T> => ({
+    get(target: T, propKey: keyof T | keyof K) {
+      return target[propKey as keyof T]
+        ? target[propKey as keyof T]
+        : (client[name as SDKKeys] as K)[propKey as keyof K];
+    },
+  });
+
+  return new Proxy(client, clientHandler);
```

Comment:
Not sure everything is correct here:
1) `SDKKeys` contains `setOneTimeSdkHeader()` and maybe some extra members (this might be ok)
2) `SDKTypes` is evaluated as `any`
This is really a hard part, was same issues for me in similar pr
<img width="643" alt="Screenshot 2019-11-28 at 22 37 41" src="https://user-images.githubusercontent.com/10908415/69832774-61fe3000-1230-11ea-8005-f1f51c5fb148.png">
<img width="245" alt="Screenshot 2019-11-28 at 22 40 34" src="https://user-images.githubusercontent.com/10908415/69832775-61fe3000-1230-11ea-9a91-9fc21d423e89.png">

Reply:
I'll review it one more time

Reply:
What I actually found that props here are correct.
<img width="923" alt="Screenshot 2019-11-29 at 13 38 54" src="https://user-images.githubusercontent.com/24567091/69878661-3af63b80-12ce-11ea-96a3-b6b124bae41f.png">

You can also check this on http://www.typescriptlang.org/play/